Superfluous stack traces logged when database does not have SSL support

**Describe the bug**
When adding a new database, even if SSL option is not enabled, Metabase will try SSL first and fallback to no SSL only when connecting through SSL does not work. That is fine, however, a long stack trace containing SSL errors when SSL is not enabled can be confusing to the admin.

This is also very annoying in Cypress as we have to scroll through pages of stack trace before being able to see the actual test log.

**Logs**
Please include javascript console and server logs around the time this bug occurred. For information about how to get these, consult our [bug troubleshooting guide](https://metabase.com/docs/latest/troubleshooting-guide/bugs.html)

<details>
  <summary>Click to expand</summary>

```
2022-03-28 14:47:38,515 ERROR driver.util :: Database connection error
java.sql.SQLNonTransientConnectionException: Could not connect to address=(host=localhost)(port=3306)(type=master) : Could not connect to localhost:3306 : PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:73)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:197)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1394)
	at org.mariadb.jdbc.internal.util.Utils.retrieveProxy(Utils.java:635)
	at org.mariadb.jdbc.MariaDbConnection.newConnection(MariaDbConnection.java:150)
	at org.mariadb.jdbc.Driver.connect(Driver.java:89)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:677)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:189)
	at clojure.java.jdbc$get_driver_connection.invokeStatic(jdbc.clj:271)
	at clojure.java.jdbc$get_driver_connection.invoke(jdbc.clj:250)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:411)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:245)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:242)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:253)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:249)
	at metabase.driver.sql_jdbc$fn__81994.invokeStatic(sql_jdbc.clj:35)
	at metabase.driver.sql_jdbc$fn__81994.invoke(sql_jdbc.clj:33)
	at metabase.driver.mysql$fn__79873.invokeStatic(mysql.clj:71)
	at metabase.driver.mysql$fn__79873.invoke(mysql.clj:67)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__19263.invoke(util.clj:48)
	at metabase.util$do_with_timeout$fn__6209.invoke(util.clj:352)
	at clojure.core$binding_conveyor_fn$fn__5772.invoke(core.clj:2034)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.sql.SQLNonTransientConnectionException: Could not connect to localhost:3306 : PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:73)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:188)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.createConnection(AbstractConnectProtocol.java:578)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1389)
	... 32 more
Caused by: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:352)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:295)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:290)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1357)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1232)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1175)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:421)
	at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:182)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:172)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1501)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1411)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:451)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:422)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.sslWrapper(AbstractConnectProtocol.java:661)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.createConnection(AbstractConnectProtocol.java:544)
	... 33 more
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:439)
	at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306)
	at java.base/sun.security.validator.Validator.validate(Validator.java:264)
	at java.base/sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:222)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:129)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1341)
	... 46 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
	at java.base/java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297)
	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:434)
	... 52 more
2022-03-28 14:47:38,517 ERROR api.database :: Cannot connect to Database
java.lang.Exception: Could not connect to address=(host=localhost)(port=3306)(type=master) : Could not connect to localhost:3306 : PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at metabase.driver.util$can_connect_with_details_QMARK_.invokeStatic(util.clj:53)
	at metabase.driver.util$can_connect_with_details_QMARK_.doInvoke(util.clj:36)
	at clojure.lang.RestFn.invoke(RestFn.java:442)
	at metabase.api.database$test_database_connection.invokeStatic(database.clj:431)
	at metabase.api.database$test_database_connection.doInvoke(database.clj:421)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at metabase.api.database$fn__73329$test_connection_details__73334$fn__73335.invoke(database.clj:478)
	at metabase.api.database$fn__73329$test_connection_details__73334.invoke(database.clj:460)
	at metabase.api.database$fn__73361.invokeStatic(database.clj:498)
	at metabase.api.database$fn__73361.invoke(database.clj:484)
	at compojure.core$wrap_response$fn__32273.invoke(core.clj:160)
	at compojure.core$wrap_route_middleware$fn__32257.invoke(core.clj:132)
	at compojure.core$wrap_route_info$fn__32262.invoke(core.clj:139)
	at compojure.core$wrap_route_matches$fn__32266.invoke(core.clj:151)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285.invoke(core.clj:200)
	at metabase.server.middleware.auth$enforce_authentication$fn__65369.invoke(auth.clj:14)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285.invoke(core.clj:200)
	at compojure.core$make_context$handler__32313.invoke(core.clj:289)
	at compojure.core$make_context$fn__32317.invoke(core.clj:299)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$make_context$fn__32317.invoke(core.clj:300)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$make_context$fn__32317.invoke(core.clj:300)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$make_context$fn__32317.invoke(core.clj:300)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$make_context$fn__32317.invoke(core.clj:300)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$make_context$fn__32317.invoke(core.clj:300)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$make_context$fn__32317.invoke(core.clj:300)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at metabase.api.routes$fn__76493$fn__76496.invoke(routes.clj:56)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285.invoke(core.clj:200)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.server.routes$fn__76638$fn__76639.doInvoke(routes.clj:57)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285.invoke(core.clj:200)
	at compojure.core$make_context$handler__32313.invoke(core.clj:289)
	at compojure.core$make_context$fn__32317.invoke(core.clj:299)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__32266.invoke(core.clj:153)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__32266.invoke(core.clj:153)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__32266.invoke(core.clj:153)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285$f__32286$respond_SINGLEQUOTE___32287.invoke(core.clj:197)
	at metabase.server.routes$fn__76626$fn__76628.invoke(routes.clj:41)
	at compojure.core$routes$fn__32285$f__32286.invoke(core.clj:198)
	at compojure.core$routes$fn__32285.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__73621.invoke(exceptions.clj:98)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__73618.invoke(exceptions.clj:86)
	at metabase.server.middleware.log$log_api_call$fn__77107$fn__77108$fn__77109.invoke(log.clj:209)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:15)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:9)
	at metabase.server.middleware.log$log_api_call$fn__77107$fn__77108.invoke(log.clj:201)
	at toucan.db$_do_with_call_counting.invokeStatic(db.clj:216)
	at toucan.db$_do_with_call_counting.invoke(db.clj:209)
	at metabase.server.middleware.log$log_api_call$fn__77107.invoke(log.clj:200)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__80460.invoke(browser_cookie.clj:38)
	at metabase.server.middleware.security$add_security_headers$fn__55229.invoke(security.clj:148)
	at metabase.server.middleware.json$wrap_json_body$fn__79615.invoke(json.clj:62)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__79633.invoke(json.clj:98)
	at metabase.server.middleware.offset_paging$handle_paging$fn__55253.invoke(offset_paging.clj:42)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__80727.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__80743.invoke(params.clj:69)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__34379.invoke(misc.clj:59)
	at metabase.server.middleware.session$bind_current_user$fn__45722$fn__45723.invoke(session.clj:262)
	at metabase.server.middleware.session$do_with_current_user.invokeStatic(session.clj:243)
	at metabase.server.middleware.session$do_with_current_user.invoke(session.clj:235)
	at metabase.server.middleware.session$bind_current_user$fn__45722.invoke(session.clj:261)
	at metabase.server.middleware.session$wrap_current_user_info$fn__45709.invoke(session.clj:221)
	at metabase.server.middleware.session$wrap_session_id$fn__45695.invoke(session.clj:167)
	at metabase.server.middleware.auth$wrap_api_key$fn__65377.invoke(auth.clj:27)
	at ring.middleware.cookies$wrap_cookies$fn__80647.invoke(cookies.clj:216)
	at metabase.server.middleware.misc$add_content_type$fn__34362.invoke(misc.clj:27)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__34387.invoke(misc.clj:76)
	at ring.middleware.gzip$wrap_gzip$fn__80689.invoke(gzip.clj:86)
	at metabase.server.middleware.misc$bind_request$fn__34390.invoke(misc.clj:93)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__80476.invoke(ssl.clj:38)
	at metabase.server$async_proxy_handler$fn__76880.invoke(server.clj:73)
	at metabase.server.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:386)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.sql.SQLNonTransientConnectionException: Could not connect to address=(host=localhost)(port=3306)(type=master) : Could not connect to localhost:3306 : PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:73)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:197)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1394)
	at org.mariadb.jdbc.internal.util.Utils.retrieveProxy(Utils.java:635)
	at org.mariadb.jdbc.MariaDbConnection.newConnection(MariaDbConnection.java:150)
	at org.mariadb.jdbc.Driver.connect(Driver.java:89)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:677)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:189)
	at clojure.java.jdbc$get_driver_connection.invokeStatic(jdbc.clj:271)
	at clojure.java.jdbc$get_driver_connection.invoke(jdbc.clj:250)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:411)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:245)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:242)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:253)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:249)
	at metabase.driver.sql_jdbc$fn__81994.invokeStatic(sql_jdbc.clj:35)
	at metabase.driver.sql_jdbc$fn__81994.invoke(sql_jdbc.clj:33)
	at metabase.driver.mysql$fn__79873.invokeStatic(mysql.clj:71)
	at metabase.driver.mysql$fn__79873.invoke(mysql.clj:67)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__19263.invoke(util.clj:48)
	at metabase.util$do_with_timeout$fn__6209.invoke(util.clj:352)
	at clojure.core$binding_conveyor_fn$fn__5772.invoke(core.clj:2034)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more
Caused by: java.sql.SQLNonTransientConnectionException: Could not connect to localhost:3306 : PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:73)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:188)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.createConnection(AbstractConnectProtocol.java:578)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1389)
	... 32 more
Caused by: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:352)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:295)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:290)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1357)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1232)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1175)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:421)
	at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:182)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:172)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1501)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1411)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:451)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:422)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.sslWrapper(AbstractConnectProtocol.java:661)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.createConnection(AbstractConnectProtocol.java:544)
	... 33 more
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:439)
	at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306)
	at java.base/sun.security.validator.Validator.validate(Validator.java:264)
	at java.base/sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:222)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:129)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1341)
	... 46 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
	at java.base/java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297)
	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:434)
	... 52 more
```
</details>

**To Reproduce**
Steps to reproduce the behavior:
1. Go to admin page and add a new database
2. Add a new MySQL database without SSL
3. Adding the database is successful, but a huge exception is logged to the console.

**Expected behavior**
The exception should only be logged if SSL connection is explicitly enabled.

**Screenshots**
N/A

**Information about your Metabase Installation:**
Metabase 0.42.3 on H2 connecting to MySQL

**Severity**
Annoying

--- a/src/metabase/api/database.clj
+++ b/src/metabase/api/database.clj
@@ -421,8 +421,9 @@
 (defn test-database-connection
   "Try out the connection details for a database and useful error message if connection fails, returns `nil` if
    connection succeeds."
-  [engine {:keys [host port] :as details}, & {:keys [invalid-response-handler]
-                                              :or   {invalid-response-handler invalid-connection-response}}]
+  [engine {:keys [host port] :as details}, & {:keys [invalid-response-handler log-exception]
+                                              :or   {invalid-response-handler invalid-connection-response
+                                                     log-exception            true}}]
   {:pre [(some? engine)]}
   (let [engine  (keyword engine)
         details (assoc details :engine engine)]
@@ -445,7 +446,8 @@
         :else
         (invalid-response-handler :db (tru "Unable to connect to database.")))
       (catch Throwable e
-        (log/error e (trs "Cannot connect to Database"))
+        (when log-exception
+          (log/error e (trs "Cannot connect to Database")))
         (invalid-response-handler :dbname (.getMessage e))))))
 
 ;; TODO - Just make `:ssl` a `feature`
@@ -465,21 +467,19 @@
   the details used to successfully connect. Otherwise returns a map with the connection error message. (This map will
   also contain the key `:valid` = `false`, which you can use to distinguish an error from valid details.)"
   [engine :- DBEngineString, details :- su/Map]
-  (if (and (supports-ssl? (keyword engine))
-           (true? (:ssl details)))
-    (let [error (test-database-connection engine details)]
-      (or error details))
-    (let [details (if (supports-ssl? (keyword engine))
-                    (assoc details :ssl true)
-                    details)]
-    ;; this loop tries connecting over ssl and non-ssl to establish a connection
-    ;; if it succeeds it returns the `details` that worked, otherwise it returns an error
-      (loop [details details]
-        (let [error (test-database-connection engine details)]
-          (if (and error
-                   (true? (:ssl details)))
-            (recur (assoc details :ssl false))
-            (or error details)))))))
+  (let [;; Try SSL first if SSL is supported and not already enabled
+        ;; If not successful or not applicable, details-with-ssl will be nil
+        details-with-ssl (assoc details :ssl true)
+        details-with-ssl (when (and (supports-ssl? (keyword engine))
+                                    (not (true? (:ssl details)))
+                                    (nil? (test-database-connection engine details-with-ssl :log-exception false)))
+                           details-with-ssl)]
+    (or
+      ;; Opportunistic SSL
+      details-with-ssl
+      ;; Try with original parameters
+      (test-database-connection engine details)
+      details)))
 
 (api/defendpoint POST "/"
   "Add a new `Database`."

--- a/src/metabase/driver/util.clj
+++ b/src/metabase/driver/util.clj
@@ -49,7 +49,6 @@
       ;; actually if we are going to `throw-exceptions` we'll rethrow the original but attempt to humanize the message
       ;; first
       (catch Throwable e
-        (log/error e (trs "Database connection error"))
         (throw (Exception. (str (driver/humanize-connection-error-message driver (.getMessage e))) e))))
     (try
       (can-connect-with-details? driver details-map :throw-exceptions)

--- a/test/metabase/api/database_test.clj
+++ b/test/metabase/api/database_test.clj
@@ -737,49 +737,41 @@
     (is (= "You don't have permissions to do that."
            (mt/user-http-request :rasta :post 403 (format "database/%d/discard_values" (mt/id)))))))
 
-
-;; For some stupid reason the *real* version of `test-database-connection` is set up to do nothing for tests. I'm
-;; guessing it's done that way so we can save invalid DBs for some silly tests. Instead of doing it the right way
-;; and using `with-redefs` to disable it in the few tests where it makes sense, we actually have to use `with-redefs`
-;; here to simulate its *normal* behavior. :unamused:
-(defn- test-database-connection [engine details]
-  (if (driver.u/can-connect-with-details? (keyword engine) details)
-    nil
-    {:valid false, :message "Error!"}))
-
 (deftest validate-database-test
   (testing "POST /api/database/validate"
-    (with-redefs [database-api/test-database-connection test-database-connection]
-      (testing "Should require superuser permissions"
-        (is (= "You don't have permissions to do that."
-               (mt/user-http-request :rasta :post 403 "database/validate"
-                                     {:details {:engine :h2, :details (:details (mt/db))}}))))
+    (testing "Should require superuser permissions"
+      (is (= "You don't have permissions to do that."
+             (mt/user-http-request :rasta :post 403 "database/validate"
+                                   {:details {:engine :h2, :details (:details (mt/db))}}))))
 
-      (testing "Underlying `test-connection-details` function should work"
-        (is (= (:details (mt/db))
-               (#'database-api/test-connection-details "h2" (:details (mt/db))))))
+    (testing "Underlying `test-connection-details` function should work"
+      (is (= (:details (mt/db))
+             (#'database-api/test-connection-details "h2" (:details (mt/db))))))
 
-      (testing "Valid database connection details"
-        (is (= {:valid true}
+    (testing "Valid database connection details"
+      (is (= {:valid true}
+             (mt/user-http-request :crowberto :post 200 "database/validate"
+                                   {:details {:engine :h2, :details (:details (mt/db))}}))))
+
+    (testing "invalid database connection details"
+      (testing "calling test-connection-details directly"
+        (is (= {:dbname  "Hmm, we couldn't connect to the database. Make sure your host and port settings are correct"
+                :message "Hmm, we couldn't connect to the database. Make sure your host and port settings are correct"
+                :valid   false}
+               (#'database-api/test-connection-details "h2" {:db "ABC"}))))
+
+      (testing "via the API endpoint"
+        (is (= {:valid false}
                (mt/user-http-request :crowberto :post 200 "database/validate"
-                                     {:details {:engine :h2, :details (:details (mt/db))}}))))
-
-      (testing "invalid database connection details"
-        (testing "calling test-connection-details directly"
-          (is (= {:valid false, :message "Error!"}
-                 (#'database-api/test-connection-details "h2" {:db "ABC"}))))
-
-        (testing "via the API endpoint"
-          (is (= {:valid false}
-                 (mt/user-http-request :crowberto :post 200 "database/validate"
-                                       {:details {:engine :h2, :details {:db "ABC"}}}))))))
+                                     {:details {:engine :h2, :details {:db "ABC"}}})))))
 
     (let [call-count (atom 0)
-          ssl-values (atom [])]
-      (with-redefs [database-api/test-database-connection (fn [_ details]
+          ssl-values (atom [])
+          valid?     (atom false)]
+      (with-redefs [database-api/test-database-connection (fn [_ details & _]
                                                             (swap! call-count inc)
                                                             (swap! ssl-values conj (:ssl details))
-                                                            {:valid true})]
+                                                            (if @valid? nil {:valid false}))]
         (testing "with SSL enabled, do not allow non-SSL connections"
           (#'database-api/test-connection-details "postgres" {:ssl true})
           (is (= 1 @call-count))
@@ -799,7 +791,17 @@
         (testing "with SSL unspecified, try twice (once with, once without SSL)"
           (#'database-api/test-connection-details "postgres" {})
           (is (= 2 @call-count))
-          (is (= [true false] @ssl-values)))))))
+          (is (= [true nil] @ssl-values)))
+
+        (reset! call-count 0)
+        (reset! ssl-values [])
+        (reset! valid? true)
+
+        (testing "with SSL disabled, but working try once (since SSL work we don't try without SSL)"
+          (is (= {:ssl true}
+                 (#'database-api/test-connection-details "postgres" {:ssl false})))
+          (is (= 1 @call-count))
+          (is (= [true] @ssl-values)))))))
 
 
 ;;; +----------------------------------------------------------------------------------------------------------------+

